src/symbols.c

```diff
@@ -1267,38 +1274,94 @@ static gboolean tree_store_remove_row(GtkTreeStore *store, GtkTreeIter *iter)
 }
 
 
-/* adds a new element in the parent table if it's key is known.
- * duplicates are kept */
-static void update_parents_table(GHashTable *table, const TMTag *tag, const gchar *parent_name,
-		const GtkTreeIter *iter)
+static gint tree_search_func(gconstpointer key, gpointer user_data)
 {
-	GList **list;
-	if (g_hash_table_lookup_extended(table, tag->name, NULL, (gpointer *) &list) &&
-		! utils_str_equal(parent_name, tag->name) /* prevent Foo::Foo from making parent = child */)
+	TreeSearchData *data = user_data;
+	gint parent_line = GPOINTER_TO_INT(key);
+	gboolean new_nearest;
+
+	if (data->found_line == -1)
+		data->found_line = parent_line; /* initial value */
+
+	new_nearest = ABS(data->line - parent_line) < ABS(data->line - data->found_line);
+
+	if (parent_line > data->line)
 	{
-		if (! list)
+		if (new_nearest && !data->lower)
+			data->found_line = parent_line;
+		return -1;
+	}
+
+	if (new_nearest)
+		data->found_line = parent_line;
+
+	if (parent_line < data->line)
+		return 1;
+
+	return 0;
+}
+
+
+static gint tree_cmp(gconstpointer a, gconstpointer b, gpointer user_data)
+{
+	return GPOINTER_TO_INT(a) - GPOINTER_TO_INT(b);
+}
+
+
+static void parents_table_tree_value_free(gpointer data)
+{
+	g_slice_free(GtkTreeIter, data);
+}
+
+
+/* adds a new element in the parent table if it's key is known. */
+static void update_parents_table(GHashTable *table, gchar *name, const gchar *parent_name,
+		GtkTreeIter *iter, GtkTreeStore *store)
+{
+	GTree *tree;
+	if (g_hash_table_lookup_extended(table, name, NULL, (gpointer *) &tree) &&
+		! utils_str_equal(parent_name, name) /* prevent Foo::Foo from making parent = child */)
+	{
+		TMTag *tag;
+
+		if (!tree)
 		{
-			list = g_slice_alloc(sizeof *list);
-			*list = NULL;
-			g_hash_table_insert(table, tag->name, list);
+			tree = g_tree_new_full(tree_cmp, NULL, NULL, parents_table_tree_value_free);
+			g_hash_table_insert(table, name, tree);
 		}
-		*list = g_list_prepend(*list, g_slice_dup(GtkTreeIter, iter));
+
+		gtk_tree_model_get(GTK_TREE_MODEL(store), iter, SYMBOLS_COLUMN_TAG, &tag, -1);
```

Comment:
I don't like looking up the store if it can be avoided.  And IIUC it can, "just" needs to also receive the old tag as a second arg: in the first pass we already looked it up, and in the second it'll be the same.